src/Nethermind/Nethermind.Consensus/Processing/CensorshipDetector/CensorshipDetector.cs

```diff
@@ -0,0 +1,187 @@
+    private readonly LruCache<BlockNumberHash, BlockCensorshipInfo> _potentiallyCensoredBlocks;
+    private IEnumerable<BlockNumberHash> _censoredBlocks = [];
+    private const int _cacheSize = 64;
+
+    public CensorshipDetector(
+        ITxPool txPool,
+        IComparer<Transaction> comparer,
+        IBlockProcessor blockProcessor,
+        ILogManager logManager,
+        ICensorshipDetectorConfig censorshipDetectorConfig)
+    {
+        _txPool = txPool;
+        _comparer = comparer;
+        _blockProcessor = blockProcessor;
+        _logger = logManager?.GetClassLogger() ?? throw new ArgumentNullException(nameof(logManager));
+        _censorshipDetectorConfig = censorshipDetectorConfig;
+
+        if (_censorshipDetectorConfig.AddressesForCensorshipDetection is not null)
+        {
+            foreach (string hexString in _censorshipDetectorConfig.AddressesForCensorshipDetection)
+            {
+                if (Address.IsValidAddress(hexString, true) && Address.TryParse(hexString, out Address address))
+                {
+                    _poolAddressCensorshipDetectorHelper.Add(address!, 0);
+                }
+            }
+        }
+
+        _potentiallyCensoredBlocks = new(_cacheSize, _cacheSize, "potentiallyCensoredBlocks");
+        _txPool.NewPending += OnAddingTxToPool;
+        _txPool.RemovedPending += OnRemovingTxFromPool;
+        _txPool.EvictedPending += OnRemovingTxFromPool;
+        _blockProcessor.BlockProcessing += OnBlockProcessing;
+    }
+
+    private void OnAddingTxToPool(object? sender, TxPool.TxEventArgs e)
+    {
+        Transaction tx = e.Transaction;
+        if (tx.To is not null && _poolAddressCensorshipDetectorHelper.TryGetValue(tx.To!, out long txSentToAddressCount))
+        {
+            if (txSentToAddressCount == 0)
+            {
+                Metrics.PoolCensorshipDetectionUniqueAddressesCount++;
+            }
+            _poolAddressCensorshipDetectorHelper[tx.To!]++;
+        }
+    }
+
+    private void OnRemovingTxFromPool(object? sender, TxPool.TxEventArgs e)
+    {
+        Transaction tx = e.Transaction;
+        if (tx.To is not null && _poolAddressCensorshipDetectorHelper.TryGetValue(tx.To!, out long txSentToAddressCount) && txSentToAddressCount > 0)
+        {
+            if (txSentToAddressCount == 1)
+            {
+                Metrics.PoolCensorshipDetectionUniqueAddressesCount--;
+            }
+            _poolAddressCensorshipDetectorHelper[tx.To!]--;
+        }
+    }
+
+    private void OnBlockProcessing(object? sender, BlockEventArgs e)
+    {
+        Task.Run(() => Cache(e.Block));
+    }
+
+    private void Cache(Block block)
+    {
+        Transaction bestTxInPool = _txPool.GetBestTx();
+        long _poolCensorshipDetectionUniqueAddressCount = Metrics.PoolCensorshipDetectionUniqueAddressesCount;
+
+        Transaction bestTxInBlock = block.Transactions[0];
+        SortedDictionary<Address, bool> _blockAddressCensorshipDetectorHelper = [];
```

Comment:
What is the point of this dictionary? If only to check that we count an address once, then `HashSet` should work better